### PR TITLE
osra405 Hq admin_user authentication

### PR DESCRIPTION
**Pull request by razorcd**
https://osraav.atlassian.net/browse/OSRA-405

- added Rails Login/Logout functionality using devise gem
- changed devise messages from using `Sign in`/`Sign out` to use `Log in`/`Log out`
- removed AA login functionality (had to)
- replaced HQ cucumber features with rspec features and moved feature methods in modules
- added login/logout route specs
- small cosmetic css changes

### CODE REVIEW

File: app/models/admin_user.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class AdminUser < ActiveRecord::Base
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
-  devise :database_authenticatable, 
-         :recoverable, :rememberable, :trackable, :validatable
+  devise :database_authenticatable, :rememberable
```
Comment (gmunumel) [inline app/models/admin_user.rb]: `:memorable` instead of `:rememberable`

File: app/views/layouts/_header.html.erb [hunk 2]
```diff
@@ -2,7 +2,7 @@
   <div class="container-fluid">
     <!-- Brand and toggle get grouped for better mobile display -->
     <div class="navbar-header">
-      <a class="navbar-brand" href="#">OSRA Support System</a>
+      <a class="navbar-brand" href="/">OSRA Support System</a>
```
Comment (gmunumel) [inline app/views/layouts/_header.html.erb]: Is this causing a real impact? With `/` is the page loaded again?

File: app/models/admin_user.rb (hunk 1 above)
Comment (razorcd) [inline app/models/admin_user.rb]: It is `:rememberable`, from devise. It is for the "remember me" checkbox at login form and saving the token. See more here: https://github.com/plataformatec/devise

File: app/views/layouts/_header.html.erb (hunk 2 above)
Comment (razorcd) [inline app/views/layouts/_header.html.erb]: It points to the root which in our case loads the AA OSRA dashboard.
I thing it's more of a convention to link the header logo to the homepage. Every public website does this.

**Comment by razorcd** [comment]
loginhqform

loginhqform2

loginhqform3

loginhqform4

**Comment by razorcd** [comment]
By default it will load the hq login form. (AA login form/routes removed) But once logged in, it will redirect to AA dashboard.

**Comment by razorcd** [comment]
Notice: because we are not showing the left navbar, the `success` flash message: `Logged out successfully.` is a little off. 
An easy fix would be to just move the flash message just between the header and navbar and remove the border and margin, so it looks more like in AA:

aa

**Comment by gmunumel** [comment]
Looks great :+1:

**Comment by razorcd** [comment]
Hq::Devise and Admin::Devise now run in parallel.   ActiveAdmin functionality is not affected.

File: app/models/admin_user.rb (hunk 1 above)
Comment (gmunumel) [inline app/models/admin_user.rb]: Yes. Sorry for that silly question :(

**Comment by gmunumel** [comment]
:shipit:

File: config/initializers/devise.rb [hunk 3]
```diff
@@ -224,7 +224,7 @@
   # config.navigational_formats = ['*/*', :html]
 
   # The default HTTP method used to sign out a resource. Default is :delete.
-  config.sign_out_via = :delete
+  config.sign_out_via = :get
```
Comment (NikitaAvvakumov) [inline config/initializers/devise.rb]: Why this change? Ideally, `GET` requests should be idempotent, while `DELETE` seems very appropriate here.

File: config/locales/devise.en.yml [hunk 4]
```diff
@@ -41,15 +41,15 @@ en:
...
+      signed_out: "Logged out successfully."
     unlocks:
       send_instructions: "You will receive an email with instructions about how to unlock your account in a few minutes."
       send_paranoid_instructions: "If your account exists, you will receive an email with instructions about how to unlock it in a few minutes."
-      unlocked: "Your account has been unlocked successfully. Please sign in to continue."
+      unlocked: "Your account has been unlocked successfully. Please log in to continue."
   errors:
     messages:
-      already_confirmed: "was already confirmed, please try signing in"
+      already_confirmed: "was already confirmed, please try loggin in"
```
Comment (NikitaAvvakumov) [inline config/locales/devise.en.yml]: '.. logging in'

File: spec/features/hq/admin_user_authentication_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  scenario 'Restricted pages should be accessible to logged in users' do
+    i_sign_in_as_admin
+    and_i_should_be_on_root_page
+  end
+
+  scenario 'Restricted pages should not be accessible to logged out users' do
+    i_sign_in_as_admin
+    i_sign_out
+    and_i_should_be_on_new_admin_user_session_page
```
Comment (NikitaAvvakumov) [inline spec/features/hq/admin_user_authentication_spec.rb]: Minor point: especially in a feature test, it's probably a good idea to use more user-oriented terms. Here, `and_i_should_be_on_sign_in_page` rather than `new_admin_user_session_page`.

File: spec/features/hq/user_is_creating_multiple_sponsors_spec.rb [hunk 6]
```diff
@@ -16,18 +16,11 @@
   end
 end
 
+
```
Comment (NikitaAvvakumov) [inline spec/features/hq/user_is_creating_multiple_sponsors_spec.rb]: ?

File: spec/features/hq/user_is_creating_multiple_sponsors_spec.rb [hunk 7]
```diff
@@ -16,18 +16,11 @@
...
 def an_agent_exists
   FactoryGirl.create :agent
 end
 
-def i_sign_in_as_admin
```
Comment (NikitaAvvakumov) [inline spec/features/hq/user_is_creating_multiple_sponsors_spec.rb]: How come this isn't needed anymore?

Comment (razorcd) [inline spec/features/hq/user_is_creating_multiple_sponsors_spec.rb]: just moved to `spec/features/hq/admin_user_authentication_spec.rb` together with all auth methods

File: config/initializers/devise.rb (hunk 3 above)
Comment (razorcd) [inline config/initializers/devise.rb]: because AA is using the `GET` for the `SignOut` link  from header.
But I prefer DELETE also

File: spec/features/hq/user_is_creating_multiple_sponsors_spec.rb (hunk 7 above)
Comment (NikitaAvvakumov) [inline spec/features/hq/user_is_creating_multiple_sponsors_spec.rb]: I'm still confused: How are these method definitions shared between files? Can the 'user_is_creating...' feature run on its own?

Comment (razorcd) [inline spec/features/hq/user_is_creating_multiple_sponsors_spec.rb]: good point. i think these methods should go in a helper module

**Comment by razorcd** [comment]
UPDATE: rspec Feature tests have helper modules for the shared methods.

**Comment by razorcd** [comment]
The only inconvenient for this authorization in rails is that you need to log in in `/admin ` to access the active admin pages and the log in in `/hq` to access the hq pages.
AA and Rails should run in parallel so I think we should keep this behavior.

File: config/routes.rb [hunk 8]
```diff
@@ -1,6 +1,9 @@
 Osra::Application.routes.draw do
 
+  devise_for :admin_users, path: "hq", as: "hq", path_names: { sign_in: 'login', sign_out: 'logout' },
+                           :controllers => { :sessions => "hq/devise/sessions" }
   namespace :hq do
+    root to: "sponsors#index"                 #should be changed to "dashboard" when implemented"
```
Comment (theB264) [inline config/routes.rb]: Instead of a comment here, I'd recommend a spec marked pending.  When it passes, it will notify you to fix it.  You can put this comment inside the routing spec for the dashboard view.

**Comment by robbiejackson** [comment]
Christian just checking up with you regarding the folder structure. For the controllers you've got devise under app/controllers/hq as app/controllers/hq/devise 
but for the views you've app/views/devise with hq in the middle between views and devise. Is this what you think is best?

File: spec/routing/hq/dashboard_routing_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+RSpec.describe "Hq::Dashboard redirecting", type: :request, pending: true do
+  before :each do
+    sign_in instance_double(AdminUser)
+  end
+
+  it "root redirects to dashboard" do
+    get '/hq'
+    expect(response).to redirect_to hq_dashboard_path
+  end
+end
```
Comment (razorcd) [inline spec/routing/hq/dashboard_routing_spec.rb]: @theB264 Yes, using pending specs looks more professional. Thanks.
I created this spec but when I run it, `sign_in instance_double(AdminUser)` is giving me an error:
`Could not find a valid mapping for #<RSpec::Mocks::InstanceVerifyingDouble:0x307cee0 @name="AdminUser (instance)">`
I did research to fix it but no luck. Do you have any idea how to make this work?

There is some info about this on Devise docs but not very helpful:
`@request.env["devise.mapping"] = Devise.mappings[:user]`

**Comment by razorcd** [comment]
@robbiejackson yes, you are right. I moved `app/views/devise` to `app/views/hq/devise`

Thanks. Good find!

File: spec/routing/hq/dashboard_routing_spec.rb (hunk 9 above)
Comment (theB264) [inline spec/routing/hq/dashboard_routing_spec.rb]: What I meant is instead of a comment in the routes file which can be forgotten or overlooked, you could put something like this in, and it will force you to edit code and read the comment once the route exists.
``` ruby
  RSpec.describe 'Hq::Dashboard redirecting', type: :routing do
    specify 'disposable spec', pending: 'be sure to route the default "/hq" route here once available' do
      expect(get: '/hq/dashboard').to be_routable
    end
  end
```

Comment (razorcd) [inline spec/routing/hq/dashboard_routing_spec.rb]: I see. This is not testing the `redirect` (/hq to /hq/dashboard) but it will remind the developer when it is working on dashboard.
